tasks/src/main/scala/dagr/tasks/picard/IlluminaBasecallsToSam.scala

```diff
@@ -0,0 +1,96 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+
+package dagr.tasks.picard
+
+import java.text.SimpleDateFormat
+
+import dagr.core.execsystem._
+import dagr.core.tasksystem.{Retry, VariableResources}
+import dagr.tasks.DagrDef.{DirPath, FilePath}
+import htsjdk.samtools.util.Iso8601Date
+import picard.util.IlluminaUtil.IlluminaAdapterPair
+
+import scala.collection.mutable.ListBuffer
+
+class IlluminaBasecallsToSam(basecallsDir: DirPath,
+                             lane: Int,
+                             runBarcode: String,
+                             readStructure: String,
+                             libraryParamsFile: FilePath,
+                             runDate: Option[Iso8601Date] = None,
+                             sequencingCenter: Option[String] = None,
+                             includeNonPfReads: Boolean = false,
+                             minThreads: Int = 4,
+                             maxThreads: Int = 16,
+                             adapterPairs: Seq[IlluminaAdapterPair] = Seq(
+                               IlluminaAdapterPair.INDEXED,
+                               IlluminaAdapterPair.DUAL_INDEXED,
+                               IlluminaAdapterPair.NEXTERA_V2,
+                               IlluminaAdapterPair.FLUIDIGM
+                             ),
+                             barcodesDir: Option[DirPath] = None,
+                             maxReadsInRamPerTile: Option[Int] = None,
+                             tmpDir: Option[DirPath] = None
+                            ) extends PicardTask with VariableResources with Retry {
+
+  private val byMemoryPerThread: Memory = Memory("250MB")
+  private var memoryPerThread: Memory = Memory("750MB")
+
+  /** Increases the memory per core/thread and returns true if we can run with the fewest # of threads. */
+  override def retry(systemResources: SystemResources, taskInfo: TaskExecutionInfo): Boolean = {
+    require(taskInfo.resources.memory == this.resources.memory, "Scheduled memory does not equal current memory")
+    // Increase the amount of memory required per-core
+    this.memoryPerThread = this.memoryPerThread + this.byMemoryPerThread
+    // See if we can run in the maximum amount of system memory with the fewest # of cores.
+    Memory(this.memoryPerThread.value * minThreads) <= systemResources.systemMemory
+  }
+
+  /** Chooses the maximum # of cores given a memory per core requirement. */
+  override def pickResources(resources: ResourceSet): Option[ResourceSet] = {
+    Range.inclusive(start=maxThreads, end=minThreads, step= -1)
+      .flatMap { cores =>
+        resources.subset(Cores(cores), Memory(cores * memoryPerThread.value))
+      }.headOption
+  }
```

Comment:
I am not convinced this is the right strategy.  I have seen many cases where we have A LOT more memory per core than 1G and it still wont succeed.  I am tempted to implement something that will increase the memory per core, as well as reducing the # of cores.